### PR TITLE
S3 source

### PR DESCRIPTION
**Pull request by sonalkr132**
# Description:

Supersedes: #1134 

> Currently s3 sources are almost supported in rubygems, but it will only work if you are lucky enough that your AWS secret key has no special characters in it. This patch fixes that by moving the config into the .gemrc.
> Instead of encoding the id and secret into the url, edit the .gemrc file with the authentication needed for each source. You must add the s3 bucket to the regular sources, then add the s3_source key with a set of credentials for each s3 hostname.
> 
> ```
> :sources:
> - s3://bucket1/path
> - s3://bucket2/
> - https://rubygems.org/
> s3_source: {
>   bucket1: {
>     id: "AOUEAOEU123123AOEUAO",
>     secret: "aodnuhtdao/saeuhto+19283oaehu/asoeu+123h"
>   },
>   bucket2: {
>     id: "AOUEAOEU123123AOEUAO",
>     secret: "aodnuhtdao/saeuhto+19283oaehu/asoeu+123h"
>   }
> }
> ```
# Tasks:
- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends
- [ ] [Squash commits](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html)

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by segiddins** [comment]
Looks good! Only thing I'd ask for is test coverage for all of the different failure paths

**Comment by sonalkr132** [comment]
Updated!

**Comment by olleolleolle** [comment]
Does this need amending any documentation?

I trawled around a little, trying to find "How do I configure my environment to use S3 as a source?"

**Comment by bronzdoc** [comment]
@sonalkr132 i think this just needs a rebase :)

**Comment by duckinator** [comment]
@sonalkr132 could you rebase this, please? Or would you like me to do it for you?

@bronzdoc @segiddins if this is rebased and tests pass after that, is there any reason to not merge it?

**Comment by sonalkr132** [comment]
done

File: lib/rubygems/remote_fetcher.rb [hunk 1]
```diff
@@ -414,4 +412,17 @@ def pools_for proxy
       @pools[proxy] ||= Gem::Request::ConnectionPools.new proxy, @cert_files
     end
   end
+
+  def s3_source_auth(uri)
+    return [uri.user, uri.password] if uri.user && uri.password
+    s3_source = Gem.configuration[:s3_source] || Gem.configuration['s3_source']
+    host = uri.host
+    raise FetchError.new('no s3_source key exists in .gemrc', "s3://#{host}") unless s3_source
```
Comment (bronzdoc) [inline lib/rubygems/remote_fetcher.rb]: could you change this to double quotes? just for consistency, thanks :)

File: lib/rubygems/remote_fetcher.rb [hunk 2]
```diff
@@ -414,4 +412,17 @@ def pools_for proxy
       @pools[proxy] ||= Gem::Request::ConnectionPools.new proxy, @cert_files
     end
   end
+
+  def s3_source_auth(uri)
```
Comment (duckinator) [inline lib/rubygems/remote_fetcher.rb]: Could you add some newlines to break this up into logical chunks? Just adding a single blank line after the early `return` and each `raise` call in this function is sufficient imo. :slightly_smiling_face:

**Comment by duckinator** [comment]
@sonalkr132 there's a comment from myself and one from @bronzdoc. once those are resolved, I'm happy with this. :+1:

**Comment by bronzdoc** [comment]
@bundlerbot +1

**Comment by bronzdoc** [comment]
@bundlerbot r+

**Comment by aergonaut** [comment]
Is there any documentation on what files to upload to S3/ what directory structure is required in order to make this work? I don't quite understand from reading the code.

**Comment by sonalkr132** [comment]
Hi @aergonaut 
Sorry about delay in response here. I wrote a guide for this https://github.com/rubygems/guides/pull/230.
> I don't quite understand from reading the code.

s3 source support has been in rubygems for a while (this PR only fixes a bug in it). [Blogs](https://simonwo.net/code/gem-server-in-s3/) and [tools](https://github.com/tmaier/docker-s3-gemserver) on this was relatively easy to find for me.

**Pull request merged**